lms/services/assignment.py

```diff
@@ -10,55 +15,206 @@ def __init__(self, db):
         self._db = db
 
     @lru_cache(maxsize=128)
-    def get(self, tool_consumer_instance_guid, resource_link_id):
-        return (
-            self._db.query(Assignment)
-            .filter_by(
-                tool_consumer_instance_guid=tool_consumer_instance_guid,
-                resource_link_id=resource_link_id,
+    def get(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
+    ):
+        """Get an assignment using using resource_link_id, ext_lti_assignment_id or both."""
+
+        if not any([resource_link_id, ext_lti_assignment_id]):
+            raise ValueError(
+                "Can't get an assignment without neither resource_link_id or ext_lti_assignment_id"
             )
-            .one_or_none()
-        )
 
-    def get_document_url(self, tool_consumer_instance_guid, resource_link_id):
-        """
-        Return the matching document URL or None.
+        # Non canvas assignments, just use `resource_link_id`
+        if resource_link_id and not ext_lti_assignment_id:
+            return self._get_by_resource_link_id(
+                tool_consumer_instance_guid, resource_link_id
+            )
 
-        Return the document URL for the assignment with the given
-        tool_consumer_instance_guid and resource_link_id, or None.
-        """
-        assignment = self.get(tool_consumer_instance_guid, resource_link_id)
+        # Creating/editing a canvas assignment, only `ext_lti_assignment_id` is available
+        if not resource_link_id and ext_lti_assignment_id:
+            return self._get_by_ext_lti_assignment(
+                tool_consumer_instance_guid, ext_lti_assignment_id
+            )
+
+        # We have both ext_lti_assignment_id and resource_link_id, canvas launch.
+        return self._get_for_canvas_launch(
+            tool_consumer_instance_guid, resource_link_id, ext_lti_assignment_id
+        )
+
+    def exists(
```

Comment:
Kept it here in the service, it might be useful down the line somewhere other than the predicates.